src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,171 @@
+            oldNode = oldNode.next;
+        }
+        Node oldPrevious = oldNode.prev;
+        oldPrevious.next = newNode;
+        oldNode.prev = newNode;
+        newNode.prev = oldPrevious;
+        newNode.next = oldNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexExist(index);
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexExist(index);
+        Node<T> node = findNodeByIndex(index);
+        T oldNode = node.value;
+        node.value = value;
+        return oldNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexExist(index);
+        Node<T> removedNode = findNodeByIndex(index);
+        unlink(removedNode);
+        return removedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (Objects.equals(current.value, object)) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (size == 0);
+    }
+
+    @Override
+    public boolean equals(T value1, T value2) {
+        if (value1 == null) {
+            return value2 == null;
+        }
+        return value1.equals(value2);
+    }
+
+    private void isIndexExist(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " does not exist");
+        }
+    }
+
+    private void unlink(Node<T> node) {
+        if (node == head) {
+            head = node.next;
+            if (head == null) {
+                tail = null;
+            } else {
+                head.prev = null;
+            }
+        } else if (node == tail) {
+            tail = node.prev;
+            tail.next = null;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        isIndexExist(index);
+        Node<T> temp;
+        if (index < size / 2) {
+            temp = head;
+            for (int i = 0; i < index; i++) {
+                temp = temp.next;
+            }
+        } else {
+            temp = tail;
+            for (int i = size - 1; i > index; i--) {
+                temp = temp.prev;
+            }
+        }
+        if (temp == null) {
+            throw new IndexOutOfBoundsException("Index error");
+        }
+        return temp;
```

Comment:
```suggestion
    private Node<T> findNodeByIndex(int index) {
        isIndexExist(index);
        Node<T> temp;
        if (index < size / 2) {
            temp = head;
            for (int i = 0; i < index; i++) {
                temp = temp.next;
            }
        } else {
            temp = tail;
            for (int i = size - 1; i > index; i--) {
                temp = temp.prev;
            }
        }
        return temp;
```
that should be covered in isIndexExist method